Spark multiversion support

sumwale opened the pull request:
## Changes proposed in this pull request

- support for multiple Spark versions from same code base
- SparkInternals interface to abstract out the internal APIs used by SnappyData/AQP layers that have
  changed between 2.1 to 2.4 with implementations for 2.1.0/2.1.1/2.3.2
- updated build to allow for "spark.connector.version" property to build smart connector for non-default
  Spark version

## Patch testing

precheckin

## ReleaseNotes.txt changes

Documentation for multiple spark version support in smart connector mode

## Other PRs 

https://github.com/SnappyDataInc/snappy-store/pull/478
https://github.com/SnappyDataInc/snappy-aqp/pull/187

Commit:
First cut for multiple Spark version support from same code base

Commit:
Merge remote-tracking branch 'origin/master' into spark-multiversion-support

Commit:
more fixes and additions

Commit:
Merge remote-tracking branch 'origin/master' into spark-multiversion-support

Commit:
more fixes for Spark 2.3.x support

Commit:
many other fixes for Spark 2.3.x support

remaining build failures = 34 with Spark 2.3.x (was > 200 originally)

Commit:
more additions for multiple version support

Commit:
fixes

Commit:
more compatibility fixes

Commit:
more fixes

Commit:
more fixes

Commit:
fixing remaining failures

Commit:
splitting build for product vs connector

- product will always use compatible Spark version but connector can use a different one
- added couple of sub-projects (core-product and aqp-product) that will always use the compatible version
  while connector build can use a different one
- cluster will depend on normal build if connector version is same else it will use core-product

Commit:
update submodule links

Commit:
Merge remote-tracking branch 'origin/master' into spark-multiversion-support

Commit:
fixing compilation issues after merge

Commit:
minor changes to names

ashishkshukla commented:
@sumwale - I was  looking into this code base and trying to make a build for spark 2.3.2.
I noticed that we  have defined a def newSnappySessionState(snappySession: SnappySession): SnappySessionState which creates an instance of SnappySessionState for the given spark version.
It seems this needs to implemented for every spark version we support, as it has direct call from SnappySession sessionState .
We have its implementation in Spark210Internals.scala  for spark 2.1.1 and spark 2.1.0 but its implementation is  missing for spark 2.3.2. Do we need to work on the implementation and we are handing in different way as I can not see its implementation in Spark232Internals.scala .

Commit:
some build changes

Commit:
more build cleanups

Commit:
moved compatibility modules inside core

some build and runtime fixes

Commit:
more build cleanups and fixes

Commit:
build fixes

Commit:
more fixes

Commit:
Merge remote-tracking branch 'origin/master' into spark-multiversion-support

Commit:
Merge remote-tracking branch 'origin/master' into spark-multiversion-support

Commit:
update gradle-scalatest and fixes for failures

Commit:
more fixes

Commit:
more fixes

Commit:
fixes

Commit:
fixes

Commit:
fix few AQP failures in precheckin

Commit:
correct jar names to new ones in SplitClusterDUnitTest

Commit:
fix JDBC pushdown driver for Spark 2.4

Commit:
fix remaining failure in AQP

Commit:
fixes for Spark 2.3/2.4 support

Commit:
more fixes and update to scalatest 3

Commit:
test build fixes

Commit:
update gradle to 5.6.4

Commit:
support for Spark 2.4.4

Commit:
more fixes for Spark 2.4.4

Commit:
refactored for multiple 2.4.x versions

Commit:
fix kafka incompatibility in tests with new kafka 2.x

Commit:
update snappy-spark to snappy/branch-2.4 (v2.4.5)

fixing compile errors with -Yno-adapted-args

Commit:
update store link

Commit:
support for latest 2.1 Spark (2.1.3) and 2.3 Spark (2.3.4)

Commit:
fixing build and runtime issues in tests

Commit:
update spark

Commit:
code reorganization

Commit:
more reorganization

Commit:
fixes and updated modules

Commit:
build fixes

Commit:
update spark module

Commit:
more fixes and reorganization

Commit:
fix code generation issues

Commit:
fixing insert failures due to InsertIntoTable -> InsertIntoDataSourceCommand moved into analysis phase

Commit:
codegeneration fixes

Commit:
couple more fixes

Commit:
fix CTAS behaviour

Commit:
more generated code and other fixes

Commit:
fixing compatibility with Spark 2.3.x

Commit:
fix duplicate SQLTabs and other test/code fixes

Commit:
more fixes and add Spark layer support for INSERT/PUT INTO TABLE(columns) ...

Commit:
more fixes for semantic changes in 2.4.x

Commit:
fixing serialization error in toJSON in recent Spark releases

Commit:
more fixes

Commit:
minor fixes

Commit:
few more fixes

Commit:
fix premature release and acquire of region lock

Commit:
change ctx.addNewFunction to internals.addFunction throughout

- use the return value of internals.addFunction to store-back into original variable
  or else use inlineToOuterClass=true
- change EncoderPlan to no longer extend LogicalRDD (due to difference in constructors)
- other misc fixes

Commit:
fix dbtable in JDBCOptions

update compatibility module builds to honour "ideaBuild"

Commit:
update spark-jobserver

Commit:
fixing initialization issues in smart connector

Commit:
update store module

Commit:
fix spark-unsafe deps

Commit:
changes to build.gradle deps

Commit:
fixes for regression issues and others

Commit:
fixes for multiple issues

Commit:
more fixes; make RETURNS optional in CREATE FUNCTION

made CREATE FUNCTION to be consistent with Spark

Commit:
more test and GUI plan fixes

- add search and explicit cleanup of broadcast exchanges at the end of query execution
  (else they would only be cleared with GC collects the reference)
- corrected GUI plan timings and cleanup the END message to deliver it reliably
  and not leave dangling SQL tasks running forever in some cases.
- other test changes for Spark 2.4.5 to fix failures and

Commit:
fix VIEW test failure (due to auto-generated alias name mismatch) and Decimal handling in prepared statements

Commit:
fixing few more remaining failures and update spark module link

Commit:
fixing remaining failures

- fix code generation issue seen in TPCH Q20
- correct SD's SQL listener to link any jobs during planning and execution phase of a query correctly
  (reintroduced SparkListenerSQLPlanExecutionStart/End and handle SparkListenerSQLExecutionStart to search
   for any existing execution from SparkListenerSQLPlanExecutionStart then mark it as active instead of creating new one)

Commit:
fixing test

Commit:
correct uncaught handler setting in Executor with for Spark 2.4

Commit:
skip dynamic cpusPerTask setting with smart connector

also fix few dunit test failures in ColumnBatchAndExternalTableDUnitTest
